plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2271,7 +2304,7 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
         IcebergTableHandle table = (IcebergTableHandle) handle;
         ConstraintExtractor.ExtractionResult extractionResult = extractTupleDomain(constraint);
         TupleDomain<IcebergColumnHandle> predicate = extractionResult.tupleDomain();
-        if (predicate.isAll()) {
+        if (predicate.isAll() && constraint.getPredicateColumns().isEmpty()) {
```

Comment:
Just to double-check: isn't it supposed to be `||` instead of `&&`?

Reply:
I think it should be `&&` instead of `||`, if we have a query `select id from T where cast(ds as integer) = 1` and `ds` is a partition column, `predicate#isAll` here is always true.